### ISSUE TITLE
Occasional/Random Stutter

### ISSUE THREAD

**Issue by OpenRift412** [dosbox-staging/dosbox-staging#655]
**Describe the bug**
Recently, for some reason, there's this occasional but very noticeable stutter in the video, audio, and MIDI. I've tried pretty much everything suggested online but it's not working. The only solution I've found is switching from OpenGL to a different output which fixes the problem. The thing is, I want to be able to use GLSL shaders (see SVN builds) with DOSBox so I can play with a CRT shader, which of course requires me to use OpenGL output to enable.

Fixes I've tried
- Audio blocksize/prebuffer
- Adjusting cycles/core
- Changing program settings from nvidia control panel
- Adjusting frameskip
- adjusting vsync
- some others I've probably forgotten
- disabling shaders

**To Reproduce**
Use the attached config and launch a game such as Doom or Quake, play for a while and look for stutters of video, MIDI, and audio simultaneously.

**Environment (please complete the following information):**
- Windows 10 20H2
- i7 9750H, RTX 2060, 16GB RAM

**Version of dosbox-staging:**
v0.75.1

**Configuration parameters:**
[dosbox_doom2.zip](https://github.com/dosbox-staging/dosbox-staging/files/5357378/dosbox_doom2.zip)

**Additional context**
It didn't always do this, and this also seems to happen with the latest vanilla SVN build of DOSBox as well.
(e.g. screenshots, other dosbox versions impacted, etc).

**Comment by dreamer** [comment]
I can't reproduce this on Linux, and haven't seen it on Windows so far, but I need to try reproducing with yout .conf settings.

In the meantime:

I see the `.conf` file lists:

`cycles = auto` - this setting is ok, but can you also test with `cycles = max 50%`?

`glshader  = C:\Program Files (x86)\DOSBox-0.74-3\SHADERS\CRT-Lottes_tweaked.glsl` - I tested CRT-Lottes shader some time ago and I was surprised to learn it regressed - at one time it was reasonably fast, but recently turned unplayably slow on my old laptop (but works ok-ish on my desktop). To remove possibility of this being the problem, let's do next round of testing with `glshader = sharp`.

`memsize  = 64` - DOS games generally don't need so much memory and it might affect caching in modern CPUs; let's tweak it back to default `memsize = 16` for testing.

In `[mixer]` section:
`rate = 48000` while SoundBlaster and other audio is set to `44100`; perhaps DOSBox internal mixer is causing the slowdowns in 0.75.1? Let's test with `rate = 44100`.

**Comment by OpenRift412** [comment]
I tried all the things you suggested and none of them yielded different results.

**Comment by kcgen** [comment]
@OpenRift412,

Curious if the linked branch's build below behaves any differently regarding the stuffering, when `cycles = 180000`:

https://github.com/dosbox-staging/dosbox-staging/suites/1312635341/artifacts/20755238

Note that this is still a work-in-progress so bugs are expected, but curious if the changes here have any impact either good or bad.

**Comment by OpenRift412** [comment]
It's hard to tell when it won't play the MIDIs. Also for some reason it won't let me change directories or view the current directory.

**Comment by kcgen** [comment]
It's base-lined against the master branch, so has the same FluidSynth support as the master-branch (although I'm making a big assumption that that's what you're referring to regarding midi issues). 

Can you elaborate on the midi issues and directory change issues with screenshots showing the problem, and conf file and/or batch file commands that aren't working?

**Comment by OpenRift412** [comment]
![image](https://user-images.githubusercontent.com/29780788/95636885-52f62a00-0a5e-11eb-8c3a-cf7df7ffa5d5.png)
I get this error when trying to view directories, I can change directories _sometimes_, but even when I do they're registered as being empty as well.

**Comment by dreamer** [comment]
What dosbox-staging version do you experience this `mount` issue in?

What's the output of `mount` command? (with no params - it will list what directories on your host machine are mounted). It's impossible to tell what is supposed to happen after `mount C ..` if we don't know what's your current working directory :)

**Comment by kcgen** [comment]
I'm unable to reproduce this issue.  @OpenRift412 , I've followed up via PM on another forum where you've mentioned this issue (only because I don't know how to send you a private message any other way).I've created a stand-alone package for you to try that includes everything needed to run it. Curious how it goes.

**Comment by OpenRift412** [comment]
> What dosbox-staging version do you experience this `mount` issue in?
> 
> What's the output of `mount` command? (with no params - it will list what directories on your host machine are mounted). It's impossible to tell what is supposed to happen after `mount C ..` if we don't know what's your current working directory :)

The issue isn't mounting, it's viewing and sometimes changing directories, and it comes from 0.76 alpha, the build that @kcgen linked earlier.

**Comment by OpenRift412** [comment]
> I'm unable to reproduce this issue. @OpenRift412 , I've followed up via PM on another forum where you've mentioned this issue (only because I don't know how to send you a private message any other way).I've created a stand-alone package for you to try that includes everything needed to run it. Curious how it goes.

I added my own settings regarding OpenGL and shaders and the issue still persists. Also, shaders won't load at all for some reason in the zip you sent me, despite configuring it to do so.

**Comment by kcgen** [comment]
@OpenRift412,

Suggest we stick to the target here: **suttering**, and then move onto other issues.

On my windows system, I unzipped the file, double clicked `dosbox.exe`, and I runs without issue:

![2020-10-09_17-45](https://user-images.githubusercontent.com/1557255/95641492-7b2e5a80-0a57-11eb-8eca-7aad9b36ae51.png)

You don't have to `cd` anywhere. Just play the game, and report if it stutters. 

If you want the screen bigger, feel free to drop this in the `dosbox.conf`:

``` ini
[sdl]
output = texturenb
windowresolution = 800x600

[render]
scaler = none
glshader = none
```

**Comment by OpenRift412** [comment]
> @OpenRift412,
> 
> Suggest we stick to the issue you reported: **suttering**.
> 
> On my Windows 10 machine, I unzipped the file I sent you, double clicked `dosbox.exe`, and it runs without issue:
> 
> ![2020-10-09_17-45](https://user-images.githubusercontent.com/1557255/95641492-7b2e5a80-0a57-11eb-8eca-7aad9b36ae51.png)
> 
> * You don't have to `cd` anywhere.
> * Please play the game and report if it stutters.
> 
> If you want the screen bigger, feel free to drop this in the `dosbox.conf`; it should have no effect on the stuttering:
> 
> ```ini
> [sdl]
> output = texturenb
> windowresolution = 800x600
> 
> [render]
> scaler = none
> glshader = none
> ```

I've already explained that using a different output, namely `texturenb` fixes the stuttering problem, but that's not the issue. The issue is that I _want_ to use OpenGL so I can use my shaders, but GL has a stuttering issue I can't seem to resolve.

**Comment by dreamer** [comment]
@OpenRift412 about shaders - they are supposed to go in directory (it's documented in help comment to `glshaders` conf setting):

    C:\Users\OpenRift412\AppData\Local\DOSBox\glshaders\

After that, you won't need to specify full path, only `glshader = CRT-Lottes_tweaked.glsl`.

Perhaps we have a bug about loading shaders using absolute paths, but it's separate from *this* issue and should go into a new bug report (so we can focus on fixing one issue).

About failed mounts - I want to know what mount paths are you actually using, because I have no idea - maybe you are mount C:\Windows as C: inside DOSBox and Windows 10 is preventing you from accessing files in there? (I have no idea until you tell me).

Overall to reproduce you bug we need to have *exact reproduction steps*: what game are you testing, where is it installed, how do you run it, etc, etc.

**Comment by kcgen** [comment]
> The only solution I've found is switching from OpenGL to a different output which fixes the problem. 

@OpenRift412, I'm very sorry for not catching that, and appreciate your patience.

Thanks for trying that build and confirming it's still affected even with OpenGL (that build paces video frames at ~exactly 60 FPS, each timed ~16.66 ms appart regardless of vsync settings).  Feel free to delete that package -- we should focus on the master branch at this point.

**Comment by OpenRift412** [comment]
> @OpenRift412 about shaders - they are supposed to go in directory (it's documented in help comment to `glshaders` conf setting):
> 
> ```
> C:\Users\OpenRift412\AppData\Local\DOSBox\glshaders\
> ```
> 
> After that, you won't need to specify full path, only `glshader = CRT-Lottes_tweaked.glsl`.
> 
> Perhaps we have a bug about loading shaders using absolute paths, but it's separate from _this_ issue and should go into a new bug report (so we can focus on fixing one issue).
> 
> About failed mounts - I want to know what mount paths are you actually using, because I have no idea - maybe you are mount C:\Windows as C: inside DOSBox and Windows 10 is preventing you from accessing files in there? (I have no idea until you tell me).
> 
> Overall to reproduce you bug we need to have _exact reproduction steps_: what game are you testing, where is it installed, how do you run it, etc, etc.
> 
> _edit_, and let's focus on testing specific build - use latest build from: https://dosbox-staging.github.io/downloads/devel/ (ATM it's `v0.76.0-alpha-814-ga01a`).

Okay, I managed to set it up properly, and once again, same results.

**Comment by kcgen** [comment]
I noticed your system specs are impressive (and I'm not saying that's an issue!).

However to help isolate the problem, do you have access to any lower-spec'd machines kicking around just to see if this affects your high-end machine specifically?

**Comment by dreamer** [comment]
@OpenRift412 We can't say if you set it up properly, because you forgot to share reproduction steps with us :)

Perhaps you are invoking the game in some way we did not anticipate.

> Overall to reproduce your bug we need to have exact reproduction steps: what game are you testing, where is it installed, how do you run it, etc, etc.

**Comment by OpenRift412** [comment]
> I noticed your system specs are impressive (and I'm not saying that's an issue!).
> 
> However to help isolate the problem, do you have access to any lower-spec'd machines kicking around just to see if this affects your high-end machine specifically?
> 
> _"I picked groceries in my 1800 HP alcohol-fueled rocket car, and noticed it back-firing in the parking lot"_ Sir.. that car's just warming up at 200 MPH :-)

Unfortunately I do not. I have a desktop back at home but it's even more beefy than the laptop I'm on now. ;)

**Comment by OpenRift412** [comment]
> @OpenRift412 We can't say if you set it up properly, because you forgot to share reproduction steps with us :)
> 
> Perhaps you are invoking the game in some way we did not anticipate.
> 
> > Overall to reproduce your bug we need to have exact reproduction steps: what game are you testing, where is it installed, how do you run it, etc, etc.

I essentially have a modified GOG installation of the game (in this case Doom II), where I've updated the conf files and dosbox version for a more optimized experience. I usually run it through the installation's provided shortcuts, (or in my case GOG Galaxy), but the parameters used in said shortcuts most likely aren't relevant to the issue, it's merely specifying the conf files to use for each launch option (singleplayer, IPX client, IPX server, etc.).

**Comment by kcgen** [comment]
Can you prepare a stand-along package of the games files, game settings as-configured by `setup.exe`, embedded `dosbox.conf` settings inside the game directory, local MIDI sound-font, and batch file that runs `doom2.exe with the all the same command line parameters you're passing in via the GOG Galaxy front-end?

I fear without this level of details, we won't get much further (given we can't reproduce it thus far).

**Comment by dreamer** [comment]
> I essentially have a modified GOG installation of the game (in this case Doom II), where I've updated the conf files and dosbox version for a more optimized experience.

@OpenRift412 thank you, this is very helpful :)

> I usually run it through the installation's provided shortcuts, (or in my case GOG Galaxy), but the parameters used in said shortcuts most likely aren't relevant to the issue, it's merely specifying the conf files to use for each launch option (singleplayer, IPX client, IPX server, etc.).

Those parameters in fact matter very much. But if it's standard GOG scheme, then I know how it looks like.

Overall for replacing dosbox in GOG games I suggest the following procedure:

1. Backup `DOSBOX` directory in game instalaltion dir created by GOG installer (not sure if Galaxy does it the same way, but I expect it does).
2. Create new empty `DOSBOX` dir and unpack dosbox-staging inside (that's why our binary is called `dosbox.exe` - so it could be easily dropped as a replacement for existing setups)
3. Test if game works using GOG's default settings (I expect it will)
4. Only after that, add *only the settings you need* to .conf file: `dosboxGAMENAME_single.conf`.

In our case step 4 will consist of tweaking output and glshaders, but let's start with making sure step (3) works.

**Comment by OpenRift412** [comment]
> * Test if game works using GOG's default settings (I expect it will) - using shortcuts created by GOG or via Galaxy or whatever.

Did as you said, same result.

**Comment by dreamer** [comment]
@OpenRift412 So your game does not work at all? I don't understand what do you mean by "same result".

**Comment by dreamer** [comment]
I have an idea - can you use OBS to record your desktop and show us what do you classify as stutters?

**Comment by OpenRift412** [comment]
> @OpenRift412 So your game does not work at all? I don't understand what do you mean by "same result".

I mean the same stutters.

**Comment by OpenRift412** [comment]
So I've found something interesting. There's a fork known as DOSBox optionals (I used it use it a while back but switched to vanilla SVN, then staging), and using similar settings, with opengl, I was able to play with seemingly zero of the aforementioned stutters. I would suggest looking at their code and seeing how they've done their implementation of OpenGL differently than everyone else.

**Comment by kcgen** [comment]
> I would suggest looking at their code

In some ways, it's like looking in a mirror :-)
That author appears to have taken plenty of our own code and merged it without commit author credit nor stating who authored it in their changelog. :-(

**Comment by dreamer** [comment]
> how they've done their implementation of OpenGL differently than everyone else.

@OpenRift412 They didn't - that version only applies patches that are publically available, they do not develop it themselves. All OpenGL-related patches in there were already applied (and siginificantly changed) in dosbox-staging.

Do you know how to use Git? You could pinpoint change that introduced regression using git bisect.

**Comment by OpenRift412** [comment]
> @OpenRift412 They didn't - that version only applies patches that are publically available, they do not develop it themselves. All OpenGL-related patches in there were already applied (and siginificantly changed) in dosbox-staging.
> 
> Do you know how to use Git? You could pinpoint change that introduced regression using git bisect.

I do not. I was merely pointing that out for you guys. I think what I meant to say is that DOSBox Optionals handles OpenGL output differently than other current forks.

**Comment by dreamer** [comment]
DOSBox Optionals uses random unfinished patches to provide *older* version of OpenGL support, based on *older* version on DOSBox SVN (latest Optionals version is based on r4338, but I don't know what versions of Optionals nor SVN you were testing - can you fill in those blanks for us?).

**Comment by OpenRift412** [comment]
> DOSBox Optionals uses random unfinished patches to provide _older_ version of OpenGL support, based on _older_ version on DOSBox SVN (latest Optionals version is based on r4338, but I don't know what versions of Optionals nor SVN you were testing - can you fill in those blanks for us?).

I'm using the latest version available on the github releases page, that being May 1. As for the SVN version, it says that it's using r4338 just like Staging does.

**Comment by Burrito78** [comment]
Don't be afraid to open a new ticket with all the info you can give!

Also please try out the latest dev build and see if the issue is maybe fixed already - there were A LOT of changes between 0.80.1 and 0.81 DEV.
